### ISSUE TITLE
Returned errors not passed in callback function (sd-bus)

### ISSUE THREAD

**Issue by npmccallum** [systemd/systemd#8108]
### Submission type

 - Bug report

### systemd version the issue has been seen with

`systemd-234-9.fc27.x86_64`

### Used distribution

> Fedora

### In case of bug report: Expected behaviour you didn't see

After calling `sd_bus_call_method_async()` and getting an error, I expected to see that error set in the `ret_error` argument to the callback function.

### In case of bug report: Unexpected behaviour you saw

The contents of `ret_error->name` were NULL. I confirmed that the following error reply was sent using `busctl monitor`:

```
ErrorName=org.bluez.Error.NotSupported  ErrorMessage="Operation is not supported"
```

### In case of bug report: Steps to reproduce the problem

1. Pick a method that returns an error.
2. Call it with `sd_bus_call_method_async().
3. Observe that an error is returned over DBus.
4. Observe that the error is NOT set in `ret_error`.

**Comment by poettering** [comment]
Oh, this is a misunderstanding... all our message handlers use the same prototype, and that prototype gets an sd_bus_error object that may be used to *return* errors from the message handler, it's not used for passing an error into the message handler. Thus, the parameter is primarily useful in method calls, but not so much in handling signals or method replies... (it's still used in that case though, to print a log message that a method reply handler failed if debug logging is turned on).

To get the error from the method reply use `sd_bus_message_get_error()`... 

And yes, is surprising and not obvious, but I was seeing some benefit in handling all messages with the same prototype, and this resulted in this behaviour.

I figure this is a documentation bug though

**Comment by poettering** [comment]
(btw, the prefixing of that parameter name with `ret_` is supposed to indicate that it's an output parameter of the function, not an input parameter. We try to do that systematically on all our functions, but there are some places where this isn't strictly followed)

**Comment by poettering** [comment]
Fix waiting in #15854
